File: app/src/main/java/com/example/currencyrate/ConverterActivity.java
package com.example.currencyrate;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.RotateAnimation;
import android.view.animation.ScaleAnimation;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;
import androidx.appcompat.app.AppCompatActivity;
import java.util.Random;

public class ConverterActivity extends AppCompatActivity {

    private EditText etAmountGive;
    private TextView tvAmountReceive, tvExchangeRateText;
    private ImageView btnSwap, btnBack;
    private double currentRate = 92.9343;
    private final Random random = new Random();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_converter);

        etAmountGive = findViewById(R.id.etAmountGive);
        tvAmountReceive = findViewById(R.id.tvAmountReceive);
        tvExchangeRateText = findViewById(R.id.tvExchangeRateText);

        btnSwap = findViewById(R.id.btnSwap);
        btnBack = findViewById(R.id.btnBack);

        btnBack.setOnClickListener(v -> finish());

        btnSwap.setOnClickListener(v -> {
            RotateAnimation rotate = new RotateAnimation(0, 180, 
                Animation.RELATIVE_TO_SELF, 0.5f, Animation.RELATIVE_TO_SELF, 0.5f);
            rotate.setDuration(300);
            v.startAnimation(rotate);
            
            // Randomly change the rate when swapping
            currentRate = 1.0 / currentRate;
            if (currentRate < 0.1) currentRate = 60.0 + random.nextDouble() * 40.0;
            
            updateExchangeRateDisplay();
            calculateConversion();
        });

        etAmountGive.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}
            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                calculateConversion();
            }
            @Override
            public void afterTextChanged(Editable s) {}
        });

        // Set up listeners for the amount buttons (100, 500, etc.)
        setupQuickAmountButtons();
        
        updateExchangeRateDisplay();
        calculateConversion();
    }

    private void setupQuickAmountButtons() {
        View root = findViewById(android.R.id.content);
        findAllButtons(root);
    }

    private void findAllButtons(View v) {
        if (v instanceof Button) {
            v.setOnClickListener(view -> {
                animateClick(view);
                String amount = ((Button) view).getText().toString();
                etAmountGive.setText(amount);
            });
        } else if (v instanceof ViewGroup) {
            ViewGroup vg = (ViewGroup) v;
            for (int i = 0; i < vg.getChildCount(); i++) {
                findAllButtons(vg.getChildAt(i));
            }
        }
    }

    private void updateExchangeRateDisplay() {
        tvExchangeRateText.setText(String.format("1 USD = %.4f RUB", currentRate));
    }

    private void calculateConversion() {
        String input = etAmountGive.getText().toString();
        if (input.isEmpty()) {
            tvAmountReceive.setText("0.00");
            return;
        }
        try {
            double amountGive = Double.parseDouble(input);
            double result = amountGive * currentRate;
            tvAmountReceive.setText(String.format("%.2f", result));
            tvAmountReceive.setAlpha(0.5f);
            tvAmountReceive.animate().alpha(1f).setDuration(200).start();
        } catch (NumberFormatException e) {
            tvAmountReceive.setText("0.00");
        }
    }

    private void animateClick(View v) {
        ScaleAnimation anim = new ScaleAnimation(1f, 0.95f, 1f, 0.95f, 
            Animation.RELATIVE_TO_SELF, 0.5f, Animation.RELATIVE_TO_SELF, 0.5f);
        anim.setDuration(100);
        anim.setRepeatMode(Animation.REVERSE);
        anim.setRepeatCount(1);
        v.startAnimation(anim);
    }
}

File: app/src/main/java/com/example/currencyrate/ui/ConverterActivity.kt
package com.example.currencyrate.ui

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.dynamicanimation.animation.SpringAnimation
import androidx.dynamicanimation.animation.SpringForce
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.example.currencyrate.data.local.AppDatabase
import com.example.currencyrate.data.local.CurrencyEntity
import com.example.currencyrate.data.remote.CbrApi
import com.example.currencyrate.data.repository.CurrencyRepository
import com.example.currencyrate.databinding.ActivityConverterBinding
import com.example.currencyrate.viewmodel.ConverterViewModel
import retrofit2.Retrofit
import retrofit2.converter.simplexml.SimpleXmlConverterFactory
import java.util.Locale

class ConverterActivity : AppCompatActivity() {

    private lateinit var binding: ActivityConverterBinding
    
    private val viewModel: ConverterViewModel by viewModels {
        object : ViewModelProvider.Factory {
            override fun <T : ViewModel> create(modelClass: Class<T>): T {
                val database = AppDatabase.getDatabase(applicationContext)
                val retrofit = Retrofit.Builder()
                    .baseUrl("https://www.cbr.ru/")
                    .addConverterFactory(SimpleXmlConverterFactory.create())
                    .build()
                val api = retrofit.create(CbrApi::class.java)
                val repository = CurrencyRepository(api, database.currencyDao())
                return ConverterViewModel(repository) as T
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityConverterBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setupUI()
        observeViewModel()
    }

    private fun setupUI() {
        binding.btnBack.setOnClickListener { finish() }

        binding.btnSwap.setOnClickListener {
            applySpringAnimation()
            viewModel.swapCurrencies()
        }

        binding.etAmountGive.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                viewModel.onInputChanged(s.toString())
            }
            override fun afterTextChanged(s: Editable?) {}
        })

        // Открытие выбора валют (передаем строковый тип)
        binding.spinnerGive.setOnClickListener {
            CurrencySelectorSheet.newInstance("GIVE").show(supportFragmentManager, "select_give")
        }
        binding.spinnerReceive.setOnClickListener {
            CurrencySelectorSheet.newInstance("RECEIVE").show(supportFragmentManager, "select_receive")
        }

        // Слушаем ответ от шторки выбора
        supportFragmentManager.setFragmentResultListener(CurrencySelectorSheet.REQUEST_KEY, this) { _, bundle ->
            val selectedCode = bundle.getString("selected_code")
            val type = bundle.getString("type")
            if (selectedCode != null && type != null) {
                viewModel.setCurrencyByCode(selectedCode, type)
            }
        }
    }

    private fun applySpringAnimation() {
        val springAnim = SpringAnimation(binding.btnSwap, SpringAnimation.ROTATION, 180f)
        springAnim.spring.stiffness = SpringForce.STIFFNESS_LOW
        springAnim.spring.dampingRatio = SpringForce.DAMPING_RATIO_MEDIUM_BOUNCY
        springAnim.start()
        
        springAnim.addEndListener { _, _, _, _ ->
            binding.btnSwap.rotation = 0f
        }
    }

    private fun observeViewModel() {
        viewModel.allCurrencies.observe(this) { currencies ->
            if (currencies.isNotEmpty()) {
                viewModel.setInitialCurrencies(currencies)
            }
        }

        viewModel.currencyGive.observe(this) { currency ->
            binding.tvCodeGive.text = currency?.code ?: ""
            binding.tvNameGive.text = currency?.name ?: ""
        }

        viewModel.currencyReceive.observe(this) { currency ->
            binding.tvCodeReceive.text = currency?.code ?: ""
            binding.tvNameReceive.text = currency?.name ?: ""
        }

        viewModel.resultAmount.observe(this) { amount ->
            binding.tvAmountReceive.text = String.format(Locale.getDefault(), "%.2f", amount)
        }

        viewModel.rateInfo.observe(this) { info ->
            binding.tvRateInfo.text = info
        }
    }
}
